server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamGlobalRetentionResolverTests.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.cluster.ClusterName;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.Map;
+
+import static org.elasticsearch.cluster.metadata.DataStreamGlobalRetentionTests.randomGlobalRetention;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+
+public class DataStreamGlobalRetentionResolverTests extends ESTestCase {
+
+    public void testOnlyGlobalRetentionMetadata() {
+        DataStreamFactoryRetention factoryRetention = randomBoolean()
+            ? DataStreamFactoryRetention.emptyFactoryRetention()
+            : randomNonNullFactoryRetention();
+        DataStreamGlobalRetentionResolver resolver = new DataStreamGlobalRetentionResolver(factoryRetention);
+        DataStreamGlobalRetention expectedGlobalRetention = randomGlobalRetention();
+        DataStreamGlobalRetention globalRetention = resolver.resolve(
+            ClusterState.builder(ClusterName.DEFAULT).customs(Map.of(DataStreamGlobalRetention.TYPE, expectedGlobalRetention)).build()
+        );
+        assertThat(globalRetention, notNullValue());
+        assertThat(globalRetention.getDefaultRetention(), equalTo(expectedGlobalRetention.getDefaultRetention()));
+        assertThat(globalRetention.getMaxRetention(), equalTo(expectedGlobalRetention.getMaxRetention()));
+    }
+
+    public void testOnlyFactoryRetentionFallback() {
```

Comment:
Should we have one or more tests that cover when only a `default` or only a `max` retention is specified in the factory? Or is that overkill you think?

Reply:
That is covered I think, it's possible with the random generation to get one of them being null.

Reply:
Yep you're correct; I misread the method below.